### ISSUE TITLE
Bughunting: mjsunit/es6/array-from on target MIPS32 BE

### ISSUE THREAD

**Issue by ilijapavlovic** [v8mips/v8mips#409]
https://v8.mips.com/bbv8/builders/Target%20-%20big%20-%20Debug%20-%20edge/builds/7159/steps/Check/logs/array-from

**Comment by ilijapavlovic** [comment]
The error is reproduced:

```
=== mjsunit/es6/array-from ===                    
$PROJECT/v8/test/mjsunit/mjsunit.js:221: Failure: expected <"𝄞"> found <"������">

Stack: Error
    at new MjsUnitAssertionError ($PROJECT/v8/test/mjsunit/mjsunit.js:31:16)
    at fail ($PROJECT/v8/test/mjsunit/mjsunit.js:221:11)
    at assertEquals ($PROJECT/v8/test/mjsunit/mjsunit.js:292:7)
    at assertArrayLikeEquals ($PROJECT/v8/test/mjsunit/es6/array-from.js:13:5)
    at testArrayFrom ($PROJECT/v8/test/mjsunit/es6/array-from.js:80:3)
    at $PROJECT/v8/test/mjsunit/es6/array-from.js:141:1
    at $PROJECT/v8/test/mjsunit/es6/array-from.js:178:3
    throw new MjsUnitAssertionError(message);
    ^
Error
    at new MjsUnitAssertionError ($PROJECT/v8/test/mjsunit/mjsunit.js:31:16)
    at fail ($PROJECT/v8/test/mjsunit/mjsunit.js:221:11)
    at assertEquals ($PROJECT/v8/test/mjsunit/mjsunit.js:292:7)
    at assertArrayLikeEquals ($PROJECT/v8/test/mjsunit/es6/array-from.js:13:5)
    at testArrayFrom ($PROJECT/v8/test/mjsunit/es6/array-from.js:80:3)
    at $PROJECT/v8/test/mjsunit/es6/array-from.js:141:1
    at $PROJECT/v8/test/mjsunit/es6/array-from.js:178:3
```

**Comment by ilijapavlovic** [comment]
This test failed for big endian variant. The test itself had the problem with handling surrogate pairs (a symbol represented with UTF16) on big endian target.

In this error case, tested symbol was musical symbol g cleaf "𝄞" which is encoded with surrogate pair `u"\xd834\xdd1e"`.

For big endian resulting sequence had bytes in reverted order `u"\xdd1e\xd834"`

The problem was in function `compiler::Node* LoadSurrogatePairInternal(...)` (defined in `builtins-string.cc`) which did not take care about endianess. Current solution was suitable only for little endian targets.

Following patch fixed the problem.

```
diff --git a/src/builtins/builtins-string.cc b/src/builtins/builtins-string.cc
index 6b14e25..5d05ec0 100644
--- a/src/builtins/builtins-string.cc
+++ b/src/builtins/builtins-string.cc
@@ -1379,8 +1379,13 @@ compiler::Node* LoadSurrogatePairInternal(CodeStubAssembler* assembler,

     switch (encoding) {
       case UnicodeEncoding::UTF16:
+#if V8_TARGET_LITTLE_ENDIAN
         var_result.Bind(assembler->WordOr(
             assembler->WordShl(trail, assembler->Int32Constant(16)), lead));
+#else
+        var_result.Bind(assembler->WordOr(
+            trail, assembler->WordShl(lead, assembler->Int32Constant(16))));
+#endif
         break;

       case UnicodeEncoding::UTF32: {
```

**Comment by ilijapavlovic** [comment]
Fixed with commit f2d278a95733d05c553d01adcc54653789f06689

```
    [builtins] fix load of surrogate pair for BE platform

    Swaping the order when loading surrogate pairs on big endian
    platform. This fixes testcase string-iterator.js on big
    endian.
```

The problem is fixed: https://v8.mips.com/bbv8/builders/Target%20-%20big%20-%20Debug%20-%20edge/builds/7311
